storage/innobase/dict/dict0dict.cc

```diff
@@ -1859,16 +1859,28 @@ void dict_table_change_id_in_cache(
     dict_table_t *table, /*!< in/out: table object already in cache */
     table_id_t new_id)   /*!< in: new id to set */
 {
+  ulint id_fold;
+
   ut_ad(table);
   ut_ad(dict_sys_mutex_own());
   ut_ad(table->magic_n == DICT_TABLE_MAGIC_N);
 
-  /* Remove the table from the hash table of id's */
+  id_fold = ut_fold_ull(new_id);
 
+  /* Look for a table with the same id: error if such exists */
+  {
+    dict_table_t *table2;
+    HASH_SEARCH(id_hash, dict_sys->table_id_hash, id_fold, dict_table_t *,
+                table2, ut_ad(table2->cached), table2->id == new_id);
+    ut_a(table2 == nullptr);
+  }
+
```

Comment:
better to move it after deletion of old id and before insertion with new id.
ie before assigning table->id = new_id

Reply:
fixed